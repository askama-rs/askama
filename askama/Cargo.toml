[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
