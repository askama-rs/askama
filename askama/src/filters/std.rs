use core::convert::Infallible;
use std::collections::HashMap;
use std::collections::hash_map::Entry;
use std::hash::Hash;
use std::rc::Rc;

/// Returns an iterator with all duplicates removed.
///
/// The sorting order is kept, no extra allocation is performed. However, to make it possible,
/// the data is wrapped inside `Rc`.
///
/// ```
/// # use askama::Template;
/// #[derive(Template)]
/// #[template(ext = "html", source = "{% for elem in example|unique %}{{ elem }},{% endfor %}")]
/// struct Example<'a> {
///     example: Vec<&'a str>,
/// }
///
/// assert_eq!(
///     Example { example: vec!["a", "b", "a", "c"] }.to_string(),
///     "a,b,c,"
/// );
/// ```
pub fn unique<T: Hash + Eq>(
    it: impl IntoIterator<Item = T>,
) -> Result<impl Iterator<Item = Rc<T>>, Infallible> {
    let mut set = HashMap::new();

    Ok(it.into_iter().filter_map(move |elem| {
        // To prevent cloning the data, we need to use `Rc`, like that we can clone `elem` as
        // key of the `HashSet` and return it.
        if let Entry::Vacant(entry) = set.entry(Rc::new(elem)) {
            Some(Rc::clone(entry.insert_entry(()).key()))
        } else {
            None
        }
    }))
}

#[cfg(test)]
mod test {
    use alloc::vec; // It's the macro, not the module.
    use alloc::vec::Vec;

    use super::*;

    #[test]
    fn test_unique() {
        assert_eq!(
            unique(["a", "b", "a", "c"]).unwrap().collect::<Vec<_>>(),
            vec![Rc::new("a"), Rc::new("b"), Rc::new("c")]
        );
        assert_eq!(
            unique([1, 1, 1, 2, 1]).unwrap().collect::<Vec<_>>(),
            vec![Rc::new(1), Rc::new(2)]
        );
        assert_eq!(
            unique("hello".chars()).unwrap().collect::<Vec<_>>(),
            vec![Rc::new('h'), Rc::new('e'), Rc::new('l'), Rc::new('o')]
        );
    }
}
