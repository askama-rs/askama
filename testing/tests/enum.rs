use std::any::type_name_of_val;
use std::fmt::{Debug, Display};

use askama::Template;

#[test]
fn test_book_example() {
    #[derive(Template, Debug)]
    #[template(
        source = "
            {%- match self -%}
                {%- when Self::Square(side) -%}      {{side}}^2
                {%- when Self::Rectangle { a, b} -%} {{a}} * {{b}}
                {%- when Self::Circle { radius } -%} pi * {{radius}}^2
            {%- endmatch -%}
        ",
        ext = "txt"
    )]
    enum AreaWithMatch {
        #[template(source = "{{self.0}}^2", ext = "txt")]
        Square(f32),
        #[template(source = "{{a}} * {{b}}", ext = "txt")]
        Rectangle { a: f32, b: f32 },
        #[template(source = "pi * {{radius}}^2", ext = "txt")]
        Circle { radius: f32 },
    }

    assert_eq!(AreaWithMatch::Square(2.0).render().unwrap(), "2^2");
    assert_eq!(
        AreaWithMatch::Rectangle { a: 1.0, b: 2.0 }
            .render()
            .unwrap(),
        "1 * 2",
    );
    assert_eq!(
        AreaWithMatch::Circle { radius: 3.0 }.render().unwrap(),
        "pi * 3^2"
    );

    #[derive(Template, Debug)]
    enum AreaPerVariant {
        #[template(source = "{{self.0}}^2", ext = "txt")]
        Square(f32),
        #[template(source = "{{a}} * {{b}}", ext = "txt")]
        Rectangle { a: f32, b: f32 },
        #[template(source = "pi * {{radius}}^2", ext = "txt")]
        Circle { radius: f32 },
    }

    assert_eq!(AreaPerVariant::Square(2.0).render().unwrap(), "2^2");
    assert_eq!(
        AreaPerVariant::Rectangle { a: 1.0, b: 2.0 }
            .render()
            .unwrap(),
        "1 * 2",
    );
    assert_eq!(
        AreaPerVariant::Circle { radius: 3.0 }.render().unwrap(),
        "pi * 3^2"
    );

    #[derive(Template, Debug)]
    #[template(
        source = "
            {%- block square -%}
                {{self.0}}^2
            {%- endblock -%}
            {%- block rectangle -%}
                {{a}} * {{b}}
            {%- endblock -%}
            {%- block circle -%}
                pi * {{radius}}^2
            {%- endblock -%}
        ",
        ext = "txt"
    )]
    enum AreaWithBlocks {
        #[template(block = "square")]
        Square(f32),
        #[template(block = "rectangle")]
        Rectangle { a: f32, b: f32 },
        #[template(block = "circle")]
        Circle { radius: f32 },
    }

    assert_eq!(AreaWithBlocks::Square(2.0).render().unwrap(), "2^2");
    assert_eq!(
        AreaWithBlocks::Rectangle { a: 1.0, b: 2.0 }
            .render()
            .unwrap(),
        "1 * 2",
    );
    assert_eq!(
        AreaWithBlocks::Circle { radius: 3.0 }.render().unwrap(),
        "pi * 3^2"
    );
}

#[test]
fn test_simple_enum() {
    #[derive(Template, Debug)]
    #[template(
        ext = "txt",
        source = "{{ self::type_name_of_val(self) }} | {{ self|fmt(\"{:?}\") }}"
    )]
    enum SimpleEnum<'a, B: Display + Debug> {
        #[template(source = "A")]
        A,
        #[template(source = "B()")]
        B(),
        #[template(source = "C({{self.0}}, {{self.1}})")]
        C(u32, u32),
        #[template(source = "D {}")]
        D {},
        #[template(source = "E { a: {{a}}, b: {{b}} }")]
        E {
            a: &'a str,
            b: B,
        },
        // uses default source with `SimpleEnum` as `Self`
        F,
        // uses default source with a synthetic type `__Askama__SimpleEnum__G` as `Self`
        #[template()]
        G,
    }

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::A;
    assert_eq!(tmpl.render().unwrap(), "A");

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::B();
    assert_eq!(tmpl.render().unwrap(), "B()");

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::C(12, 34);
    assert_eq!(tmpl.render().unwrap(), "C(12, 34)");

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::C(12, 34);
    assert_eq!(tmpl.render().unwrap(), "C(12, 34)");

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::D {};
    assert_eq!(tmpl.render().unwrap(), "D {}");

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::E { a: "hello", b: X };
    assert_eq!(tmpl.render().unwrap(), "E { a: hello, b: X }");

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::F;
    assert_eq!(
        tmpl.render().unwrap(),
        "&enum::test_simple_enum::SimpleEnum<enum::X> | F",
    );

    let tmpl: SimpleEnum<'_, X> = SimpleEnum::G;
    assert_eq!(
        tmpl.render().unwrap(),
        "&enum::test_simple_enum::_::__Askama__SimpleEnum__G<enum::X> | \
        __Askama__SimpleEnum__G(\
            PhantomData<&enum::test_simple_enum::SimpleEnum<enum::X>>\
        )",
    );
}

#[test]
fn test_enum_blocks() {
    #[derive(Template, Debug)]
    #[template(
        ext = "txt",
        source = "\
            {% block a -%} <a = {{ a }}> {%- endblock %}
            {% block b -%} <b = {{ b }}> {%- endblock %}
            {% block c -%} <c = {{ c }}> {%- endblock %}
            {% block d -%} <d = {{ self::type_name_of_val(self) }}> {%- endblock %}
        "
    )]
    enum BlockEnum<'a, C: Display> {
        #[template(block = "a")]
        A { a: u32 },
        #[template(block = "b")]
        B { b: &'a str },
        #[template(block = "c")]
        C { c: C },
        #[template(block = "d")]
        D,
    }

    let tmpl: BlockEnum<'_, X> = BlockEnum::A { a: 42 };
    assert_eq!(tmpl.render().unwrap(), "<a = 42>");

    let tmpl: BlockEnum<'_, X> = BlockEnum::B { b: "second letter" };
    assert_eq!(tmpl.render().unwrap(), "<b = second letter>");

    let tmpl: BlockEnum<'_, X> = BlockEnum::C { c: X };
    assert_eq!(tmpl.render().unwrap(), "<c = X>");

    assert_eq!(
        BlockEnum::<'_, X>::D.render().unwrap(),
        "<d = &enum::test_enum_blocks::_::__Askama__BlockEnum__D<enum::X>>"
    );
}

#[test]
fn associated_constants() {
    #[derive(Template, Debug)]
    #[template(
        ext = "txt",
        source = "\
            {% block a -%} {{ Self::CONST_A }} {{ self.0 }} {%- endblock %}
            {% block b -%} {{ Self::CONST_B }} {{ self.0 }} {%- endblock %}
            {% block c -%} {{ Self::func_c(self.0) }} {{ self.0 }} {%- endblock %}
        "
    )]
    enum BlockEnum<'a, T: Display> {
        #[template(block = "a")]
        A(&'a str),
        #[template(block = "b")]
        B(T),
        #[template(block = "c")]
        C(&'a T),
    }

    impl<'a, T: Display> BlockEnum<'a, T> {
        const CONST_A: &'static str = "<A>";
        const CONST_B: &'static str = "<B>";

        fn func_c(_: &'a T) -> &'static str {
            "<C>"
        }
    }

    let tmpl: BlockEnum<'_, X> = BlockEnum::A("hello");
    assert_eq!(tmpl.render().unwrap(), "<A> hello");

    let tmpl: BlockEnum<'_, X> = BlockEnum::B(X);
    assert_eq!(tmpl.render().unwrap(), "<B> X");

    let tmpl: BlockEnum<'_, X> = BlockEnum::C(&X);
    assert_eq!(tmpl.render().unwrap(), "<C> X");
}

#[derive(Debug)]
struct X;

impl Display for X {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("X")
    }
}
