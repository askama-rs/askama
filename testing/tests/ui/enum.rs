use askama::Template;

#[derive(Template)]
enum CratePathOnVariant {
    #[template(ext = "txt", source = "🫨", askama = askama)]
    Variant,
}

#[derive(Template)]
enum CratePathOnVariants {
    #[template(ext = "txt", source = "🫏", askama = askama)]
    Variant1,
    #[template(ext = "txt", source = "🪿", askama = askama)]
    Variant2,
}

#[derive(Template)]
#[template(ext = "txt", source = "🪼", askama = askama)]
enum CratePathOnBoth {
    #[template(ext = "txt", source = "🪻", askama = askama)]
    Variant,
}

#[derive(Template)]
#[template(ext = "txt", source = "🫛", askama = askama)]
enum CratePathOnAll {
    #[template(ext = "txt", source = "🫠", askama = askama)]
    Variant1,
    #[template(ext = "txt", source = "🧌", askama = askama)]
    Variant2,
}

#[derive(Template)]
#[template(
    ext = "txt",
    source = "
        {%- block a -%} a {%- endblock -%}
        {%- block b -%} b {%- endblock -%}
        {#- no block c -#}
        {%- block d -%} d {%- endblock -%}
    ",
)]
enum MissingBlockName {
    #[template(block = "a")]
    A,
    #[template(block = "b")]
    B,
    #[template(block = "c")]
    C,
    #[template(block = "d")]
    D,
}

fn main() {}
