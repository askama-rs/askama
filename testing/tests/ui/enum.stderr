error: template attribute `askama` can only be used on the `enum`, not its variants
 --> tests/ui/enum.rs:5:43
  |
5 |     #[template(ext = "txt", source = "🫨", askama = askama)]
  |                                            ^^^^^^

error: template attribute `askama` can only be used on the `enum`, not its variants
  --> tests/ui/enum.rs:11:43
   |
11 |     #[template(ext = "txt", source = "🫏", askama = askama)]
   |                                            ^^^^^^

error: template attribute `askama` can only be used on the `enum`, not its variants
  --> tests/ui/enum.rs:20:43
   |
20 |     #[template(ext = "txt", source = "🪻", askama = askama)]
   |                                            ^^^^^^

error: template attribute `askama` can only be used on the `enum`, not its variants
  --> tests/ui/enum.rs:27:43
   |
27 |     #[template(ext = "txt", source = "🫠", askama = askama)]
   |                                            ^^^^^^

error: cannot find block `c`
  --> tests/ui/enum.rs:48:24
   |
48 |     #[template(block = "c")]
   |                        ^^^
